### PR TITLE
feat/shutdown-manager

### PR DESCRIPTION
**Pull request by roivaz**
This PR add graceful termination for envoy containers, with connection draining of listeners.

The shutdown manager can be enabled for EnvoyDeployment resources using:

```yaml
spec:
  shutdownManager: {}
```

The shutdown manager can be enabled for envoy injected sidecars using the following annotation in Pods:

```yaml
metadata:
  annotations:
    marin3r.3scale.net/shutdown-manager.enabled: "true"
```

The shutdown manager is a new command in the Marin3r image that runs a small server and is deployed as a sidecar container to the Envoy container. Container lifecycle hooks are used in both the shutdown manager container and the Envoy container to orchestrate graceful shutdown of the Envoy server, waiting until all the listeners are drained or the 300s timeout is reached. 

An example of how the Envoy and the shutdown manager containers are configured:

```yaml
  containers:
    - name: envoy 
      args:
        - '-c'
        - /etc/envoy/bootstrap/config.json
        - '--service-node'
        - example
        - '--service-cluster'
        - example
        - '--component-log-level'
        - 'config:debug'
      command:
        - envoy
      image: 'envoyproxy/envoy:v1.16.0'
      lifecycle:
        preStop:
          httpGet:
            path: /shutdown
            port: 8090
            scheme: HTTP
      # rest of the container config is omitted
    - name: envoy-shtdn-mgr 
      args:
        - shutdown-manager
        - '--port'
        - '8090'
      image: 'quay.io/3scale/marin3r:v0.8.0-alpha.8'
      lifecycle:
        preStop:
          httpGet:
            path: /drain
            port: 8090
            scheme: HTTP
      # rest of the container config is omitted
```

/kind feature
/priority important-soon

### CODE REVIEW

**Comment by roivaz** [comment]
@slopezz @raelga this can now be reviewed

File: pkg/envoy/container/defaults/defaults.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	// deployment specific defaults
+	DeploymentContainerName        string = "envoy"
+	DeploymentBootstrapConfigMapV2 string = "envoy-bootstrap"
+	DeploymentBootstrapConfigMapV3 string = "envoy-bootstrap-v3"
+	DeploymentConfigVolume         string = "envoy-bootstrap"
+	DeploymentTLSVolume            string = "envoy-tls"
+	DeploymentClientCertificate    string = "envoy-client-cert"
+
+	// shutdown manager defaults
+	ShtdnMgrDefaultImageRegistry      string = "quay.io/3scale/marin3r"
```
Comment (slopezz) [inline pkg/envoy/container/defaults/defaults.go:52]: So the default shutdown-manager imager is the operator image **without specific tag**? I don't know why I thought it was envoy's image

**Comment by slopezz** [comment]
@roivaz can you update the issue description with the final CRD field and annotations naming for the shutdown-manager component? just to make sure if anyone comes in the future to check this PR issue, to not have a drift between initial and final var names...

File: pkg/envoy/container/defaults/defaults.go (hunk 1 above)
Comment (roivaz) [reply pkg/envoy/container/defaults/defaults.go:52]: No, that is just the repository where the image is. The tag of the image is calculated depending on marin3r's version:

https://github.com/3scale-ops/marin3r/blob/a6d531ba4c44b5c27b2e157d8cc89eea6501cf85/pkg/envoy/container/defaults/defaults.go#L66-L68

**Comment by roivaz** [comment]
Updated @slopezz !

File: pkg/envoy/container/defaults/defaults.go (hunk 1 above)
Comment (slopezz) [reply pkg/envoy/container/defaults/defaults.go:52]: Ok nice! didnt saw it :P

**Comment by slopezz** [comment]
> Updated @slopezz !

Thanks!

Awesome addition to graceful terminate connections!

/lgtm

**Review by raelga — APPROVED** [review]
/lgtm

Awesome as always 💯

**Pull request merged**